consensus/wal_generator.go

```diff
@@ -83,7 +83,7 @@ func WALGenerateNBlocks(t *testing.T, wr io.Writer, numBlocks int) (err error) {
 	consensusState := NewState(config.Consensus, state.Copy(), blockExec, blockStore, mempool, evpool)
 	consensusState.SetLogger(logger)
 	consensusState.SetEventBus(eventBus)
-	if privValidator != nil {
+	if privValidator != nil && config.Mode == cfg.ModeValidator {
```

Comment:
I am wary of adding these sorts of mode checks around the code base, because it can get really hard to reason about the interactions of modes with other settings. It might simplify things if the modes simply modified the configuration before starting the node, e.g. such that `ModeFullNode` would clear out any privval settings.

What do you think @melekes?

Reply:
> I am wary of adding these sorts of mode checks around the code base, because it can get really hard to reason about the interactions of modes with other settings.

++

maybe we can have a `NopValidator` (https://en.wikipedia.org/wiki/Null_object_pattern) for fullnode and seednode modes?

Reply:
Not a bad idea, although personally I am fine with just using `nil` for "no validator" as well.

Reply:
I was trying just checking nil for "no validator", but there were cases `privValidator != nil` even though privValidator is empty with `ModeFullNode`, [(case)](https://github.com/tendermint/tendermint/pull/5212/files#diff-380e79adccbaf1276bf6c91cc551cd7fR419), maybe interface typecasting issue 

so originally I added the empty() function to the interface `PrivValidator` for checking "no validator"

ex) 

```
func (pv *FilePV) Empty() bool {
    return pv == (*FilePV)(nil)
}
```    

but then changed it simply to using config because there were too many relevant correction codes.